src/other/index.jl

```diff
@@ -208,3 +211,105 @@ function add_names(ind::Index, add_ind::Index; makeunique::Bool=false)
 
     return u
 end
+
+@inline parentcols(ind::Index) = Base.OneTo(length(ind))
+@inline parentcols(ind::Index, cols) = cols
+
+### SubIndex of Index. Used by SubDataFrame, DataFrameRow, and DataFrameRows
+
+# We allow I to be AbstractIndex, to allow for extensions
+# In DataFrames.jl I is always Index
+
+struct SubIndex{I<:AbstractIndex,S<:AbstractVector{Int},T<:AbstractVector{Int}} <: AbstractIndex
+    parent::I
+    cols::S # columns from idx selected in SubIndex
+    remap::T # reverse of cols
+end
+
+SubIndex(parent::AbstractIndex, ::Colon) = parent
+
+@inline parentcols(ind::SubIndex) = ind.cols
+
+Base.@propagate_inbounds parentcols(ind::SubIndex, idx::Union{Integer, AbstractVector{<:Integer}}) =
+    ind.cols[idx]
+
+Base.@propagate_inbounds function parentcols(ind::SubIndex, idx::Symbol)
+    parentcol = ind.parent[idx]
+    @boundscheck begin
+        remap = ind.remap
+        length(remap) == 0 && lazyremap!(ind)
+        remap[parentcol] == 0 && throw(KeyError("$idx not found"))
+    end
+    return parentcol
+end
+
+Base.@propagate_inbounds parentcols(ind::SubIndex, idx::AbstractVector{Symbol}) =
+    [parentcols(ind, i) for i in idx]
+
+Base.@propagate_inbounds parentcols(ind::SubIndex, ::Colon) = ind.cols
+
+Base.@propagate_inbounds function SubIndex(parent::AbstractIndex, cols::AbstractUnitRange{Int})
+    l = last(cols)
+    f = first(cols)
+    @boundscheck if !checkindex(Bool, Base.OneTo(length(parent)), cols)
+        throw(BoundsError("invalid columns $cols selected"))
+    end
+    remap = (1:l) .- f .+ 1
+    SubIndex(parent, cols, remap)
+end
+
+Base.@propagate_inbounds function SubIndex(parent::AbstractIndex, cols::AbstractVector{Int})
+    ncols = length(parent)
+    @boundscheck if !all(x -> 0 < x ≤ ncols, cols)
+        throw(BoundsError("invalid columns $cols selected"))
+    end
+    remap = Int[]
+    SubIndex(parent, cols, remap)
+end
+
+@inline SubIndex(parent::AbstractIndex, cols::ColumnIndex) =
+    throw(ArgumentError("cols argument must be a vector (got $cols)"))
+
+Base.@propagate_inbounds SubIndex(parent::AbstractIndex, cols) =
+    SubIndex(parent, parent[cols])
+
+# a helper function that lazily creates remap when needed
+# it assumes that remap is Vector{Int} unless it is an AbstractUnitRange
```

Comment:
Is this assumption really made? We just need `remap` to be mutable and be able to hold `Int` values. So I don't think this comment is needed.

Reply:
OK. Removed